src/auth/users.repository.ts

```diff
@@ -0,0 +1,56 @@
+import { Column, Entity, PrimaryGeneratedColumn, Repository } from 'typeorm';
+import { CustomRepository } from '../database/typeorm-ex.decorator';
+import { IsString, Matches, MaxLength, MinLength } from 'class-validator';
+import {
+  ConflictException,
+  InternalServerErrorException,
+} from '@nestjs/common';
+import * as bcrypt from 'bcrypt';
+
+@Entity()
+export class UserEntity {
+  @PrimaryGeneratedColumn('uuid')
+  id: string;
+  @Column({ unique: true })
+  username: string;
+  @Column()
+  password: string;
+}
+
+export class AuthCredentialsDto {
+  @IsString()
+  @MinLength(4)
+  @MaxLength(20)
+  username: string;
+
+  @IsString()
+  @MinLength(8)
+  @MaxLength(32)
+  @Matches(/((?=.*\d)|(?=.*\W+))(?![.\n])(?=.*[A-Z])(?=.*[a-z]).*$/, {
+    message: 'Password not correct',
+  })
+  password: string;
+}
+
+// @EntityRepository is deprecated, see module description
+@CustomRepository(UserEntity)
+export class UsersRepository extends Repository<UserEntity> {
+  async createUser(credentials: AuthCredentialsDto): Promise<void> {
+    const { username, password } = credentials;
+
+    const salt = await bcrypt.genSalt();
+    const hashedPassword = await bcrypt.hash(password, salt);
+
+    const entity = this.create({ username, password: hashedPassword });
+
+    try {
+      await this.save(entity);
+    } catch (err) {
+      if (err.code === '23505') {
+        throw new ConflictException('User name already exist');
+      } else {
+        throw new InternalServerErrorException();
+      }
+    }
```

Comment:
try/catch block for errors is considered as Users business logic, so it should be in the UsersService method separately, so when you will write unit tests, you will be able to simply mock the repository and test the logic of the endpoint without accessing the database